### PR TITLE
Issue/1435 send kill timeout

### LINKED ISSUE #1435: Stop job after timeout
If the job takes too long to run, it should stop. Integrate with pkconfig so can pass environment vars.

### PR DESCRIPTION
**Pull request by e-carlin**
Fix #1435 

@robnagler I've left comments below on areas that I think need discussion.

### CODE REVIEW

File: sirepo/job_supervisor.py [hunk 1]
```diff
@@ -327,14 +321,24 @@ def _ops_to_cancel():
...
-            if op and op in self.ops:
+            if op and op in self.ops and op not in o:
                 o.append(op)
             return o
 
         r = PKDict(state=job.CANCELED)
         if (
-            not self._req_is_valid(req)
-            or self.db.status not in _RUNNING_PENDING
+# TODO(e-carlin): the 'and not op' check is not quite right. The problem is if
```
Comment (e-carlin) [inline sirepo/job_supervisor.py]: This could lead to strange UX. Read the code comment for context.

File: tests/job_timeout_test.py [hunk 2]
```diff
@@ -53,3 +54,24 @@ def test_srw(fc):
...
+        'downloadDataFile',
+        PKDict(
+            simulation_type=d.simulationType,
+            simulation_id=d.models.simulation.simulationId,
+            model='heightWeightReport',
+            frame='-1',
+            suffix='sr_long_analysis',
+        ),
+    )
+    # TODO(e-carlin): This should be something more helpful than a 404
```
Comment (e-carlin) [inline tests/job_timeout_test.py]: What's the best option here? 404 seems "fine" but not the most helpful.

**Comment by e-carlin** [comment]
@robnagler the GUI interprets 'canceled' for sequential runs and analyses as bad data and re-runs the simulation/analyses up to 3x. [This is the code](https://github.com/radiasoft/sirepo/blob/32a8a83756692a1296cae613c876c843d2cc72c8/sirepo/package_data/static/js/sirepo-plotting.js#L193) that does the retry. Ultimately the user will see an error but it takes 3 retries before it appears. Maybe this is the UX we want, or maybe not, I'm not sure.

File: tests/job_timeout_test.py (hunk 2 above)
Comment (robnagler) [reply tests/job_timeout_test.py]: I think this is a very rare case so 404 is fine.

File: sirepo/job_supervisor.py [hunk 3]
```diff
@@ -327,14 +321,24 @@ def _ops_to_cancel():
...
+# we are calling api_runCancel from run_timeout and the op that timed out was
+# an analysis op that is not normally checked for request validity. Ex
+# api_downloadDataFile does not check for a valid request because a data file
+# request is not valid (it has not computeJobHash for example). But we still
+# need to cancel downloadDataFile ops. We know we are called from run_timeout
+# if op is not None so that is how we are getting around the problem. This
+# could result in odd UX. Ex a running simulation being cancelled due to a
+# downloadDataFile request timeout in another browser window (only the
+# computJids must match between the two requests).
```
Comment (robnagler) [inline sirepo/job_supervisor.py]: Getting complicated. Let's talk about this in person. I want to understand completely.

File: sirepo/pkcli/job_agent.py [hunk 4]
```diff
@@ -238,15 +238,17 @@ def terminate(self):
...
 
-    async def _cmd(self, msg):
+    async def _cmd(self, msg, **kwargs):
         if msg.opName == job.OP_ANALYSIS and msg.jobCmd != 'fastcgi':
             return await self._fastcgi_op(msg)
         c = _Cmd
         if msg.jobRunMode == job.SBATCH:
             c = _SbatchRun if msg.isParallel else _SbatchCmd
-        p = c(msg=msg, dispatcher=self)
         if msg.jobCmd == 'fastcgi':
```
Comment (robnagler) [inline sirepo/pkcli/job_agent.py]: should this be elif?

File: sirepo/pkcli/job_agent.py [hunk 5]
```diff
@@ -273,26 +275,25 @@ def _fastcgi_accept(self, connection, *args, **kwargs):
...
-        self._fastcgi_msg_q = None
-        self._fastcgi_cmd.destroy()
-        self._fastcgi_cmd = None
-        if msg:
-            await _reply_error(msg)
-        while q.qsize() > 0:
-            await _reply_error(q.get_nowait())
-            q.task_done()
+        # If self.fastcgi_cmd is None we initiated the kill so not an error
+        if self.fastcgi_cmd:
```
Comment (robnagler) [inline sirepo/pkcli/job_agent.py]:
```py
if not self.fastcgi_cmd:
     return
```

I think we still want to log. 

Maybe put the logging back at the `except` and only call the handler if fastcgi_cmd?

File: sirepo/pkcli/job_agent.py (hunk 4 above)
Comment (e-carlin) [reply sirepo/pkcli/job_agent.py]: Yes.

File: sirepo/pkcli/job_agent.py (hunk 5 above)
Comment (e-carlin) [reply sirepo/pkcli/job_agent.py]: Ya that's clearer.

**Pull request merged**